services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/ClasspathSchemaCreator.java

```diff
@@ -0,0 +1,106 @@
+import io.vertx.core.Vertx;
+import io.vertx.ext.jdbc.JDBCClient;
+
+/**
+ * Create the expected database schema if it does not exist from SQL script bundled with the classpath.
+ */
+public class ClasspathSchemaCreator implements org.eclipse.hono.deviceregistry.jdbc.SchemaCreator {
+
+    private static final Logger log = LoggerFactory.getLogger(ClasspathSchemaCreator.class);
+
+    private final JdbcProperties deviceProperties;
+    private final JdbcProperties tenantProperties;
+    private final Vertx vertx;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param devicesProperties The configuration properties for the device store.
+     * @param tenantsProperties The configuration properties for the tenant store - may be {@code null}.
+     * @throws NullPointerException if vertx or devicesProperties is {@code null}.
+     */
+    public ClasspathSchemaCreator(final Vertx vertx, final JdbcProperties devicesProperties,
+            final JdbcProperties tenantsProperties) {
+
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(devicesProperties);
+
+        this.vertx = vertx;
+        this.deviceProperties = devicesProperties;
+        this.tenantProperties = tenantsProperties;
+    }
+
+    @Override
+    public Future<Void> createDbSchema() {
+        return createTenantSchema().compose(v -> createDeviceSchema());
+    }
+
+    private Future<Void> createTenantSchema() {
+        if (tenantProperties == null) {
+            return Future.succeededFuture();
+        }
+
+        final JDBCClient jdbcClient = JdbcProperties.dataSource(vertx, tenantProperties);
+
+        switch (SQL.getDatabaseDialect(tenantProperties.getUrl())) {
+        case SQL.DIALECT_POSTGRESQL:
+            return readScript(jdbcClient, "'classpath:/sql/postgresql/01-create.tenants.sql'");
+        case SQL.DIALECT_H2:
+            return readScript(jdbcClient, "'classpath:/sql/h2/01-create.tenants.sql'");
+        default:
+            return Future.failedFuture(new IllegalArgumentException(
+                    "Unknown database dialect '" + SQL.getDatabaseDialect(tenantProperties.getUrl()) + "'"));
+        }
+    }
+
+    private Future<Void> createDeviceSchema() {
+        final JDBCClient jdbcClient = JdbcProperties.dataSource(vertx, deviceProperties);
+
+        switch (SQL.getDatabaseDialect(deviceProperties.getUrl())) {
+        case SQL.DIALECT_POSTGRESQL:
+            return readScript(jdbcClient, "'classpath:/sql/postgresql/02-create.devices.sql'");
+        case SQL.DIALECT_H2:
+            return readScript(jdbcClient, "'classpath:/sql/h2/02-create.devices.sql'");
+        default:
+            return Future.failedFuture(new IllegalArgumentException(
+                    "Unknown database dialect '" + SQL.getDatabaseDialect(deviceProperties.getUrl()) + "'"));
+        }
+    }
+
+    private Future<Void> readScript(final JDBCClient client, final String scriptPath) {
```

Comment:
this should probably be called *executeScript* or *runScript* instead, shouldn't it?